File: src/solver.cpp
#include <cstdio>

#include "solver.hpp"

/**
 * Public
 */

bool Solver::solveSudoku(int matrix[9][9]) {
    return Solver::solvePosition(matrix, 0, 0);
}

void Solver::displaySudoku(int matrix[9][9]) {
    for (int i = 0; i < 9; i++) {
        for (int k = 0; k < 9; k++) {
            printf("%d ", matrix[i][k]);
        }

        printf("\n");
    }
}

/**
 * Private
 */

bool Solver::isValidValue(int matrix[9][9], int row, int col, int value) {
    /* Check vertically and horizontally */
    for (int i = 0; i < 9; i++) {
        if (matrix[i][col] == value || matrix[row][i] == value) {
            return false;
        }
    }

    for (int i = row; i < (row + 3); i++) {
        for (int k = col; k < (col + 3); k++) {
            if (matrix[i][k] == value) {
                return false;
            }
        }
    }

    return true;
}

bool Solver::solveNextPosition(int matrix[9][9], int row, int col) {
    if (col >= 8) {
        return Solver::solvePosition(matrix, (row + 1), 0);
    } else {
        return Solver::solvePosition(matrix, row, (col + 1));
    }
}

bool Solver::solvePosition(int matrix[9][9], int row, int col) {
    /* Sudoku is solved if trying to access row that is out of range */
    if (row >= 9) {
        return true;
    }

    /* If field is not empty, solve the next one */
    if (matrix[row][col]) {
        return Solver::solveNextPosition(matrix, row, col);
    }

    /* Try every value for each field and see if it leads to a solution */
    for (int value = 1; value < 10; value++) {
        if (Solver::isValidValue(matrix, row, col, value)) {
            matrix[row][col] = value;

            if (Solver::solveNextPosition(matrix, row, col)) {
                return true;
            }

            matrix[row][col] = 0;
        }
    }

    return false;
}

File: solvedoku/solver.py


class Solver:
    __sudoku: list = []

    @staticmethod
    def __validate_matrix(matrix: list) -> bool:
        return (len(matrix) is 9 and (len(matrix[0])) is 9)

    @staticmethod
    def __is_valid_value(row: int, col: int, value: int) -> bool:
        for i in range(0, 9):
            if Solver.__sudoku[row][i] is value or Solver.__sudoku[i][col] is value:
                return False

        block_row: int = int(row / 3) * 3
        block_col: int = int(col / 3) * 3

        for i in range(block_row, (block_row + 3)):
            for k in range(block_col, (block_col + 3)):
                if Solver.__sudoku[i][k] is value:
                    return False

        return True

    @staticmethod
    def __solve_next_position(row: int, col: int) -> bool:
        if col >= 8:
            return Solver.__solve_position((row + 1), 0)
        else:
            return Solver.__solve_position(row, (col + 1))

    @staticmethod
    def __solve_position(row: int, col: int) -> bool:
        if row >= 9:
            return True

        if Solver.__sudoku[row][col]:
            return Solver.__solve_next_position(row, col)

        for value in range(1, 10):
            if Solver.__is_valid_value(row, col, value):
                Solver.__sudoku[row][col]: int = value

                if Solver.__solve_next_position(row, col):
                    return True

                Solver.__sudoku[row][col]: int = 0

        return False

    @staticmethod
    def display(matrix: list):
        for row in matrix:
            print(row)

    @staticmethod
    def solve(matrix: list) -> list:
        if Solver.__validate_matrix(matrix):
            Solver.__sudoku: list = matrix

            if (Solver.__solve_position(0, 0)):
                return Solver.__sudoku

        return None
